### PR TITLE
WIP: add tox & pytest, checking of examples / results assertions, and fix issues

### PR DESCRIPTION
**Pull request by xmo-odoo**
brain dump so I don't forget everything I've found.

The first commit converts the bash test suite to pytest and adds tox support, it works out of the box though I also added precommit as an env and *that* doesn't work.

To test: given all the relevant python are installed on the machine (using e.g/ pyenv), can just run all the tests under all the pythons using `tox`, or run a single suite using `tox -e py38` (for instance). Running `pytest` directly should work fine.

The second commit adds #49 actual, matching the real output of `flake8` against a "known good" output, the following commits fix the various issues, the last commit (hopefully) fixes the offset issue from #38 found in #49.

This is a very raw PR:

* flake8 really wants every pytest test to have its own docstring and I could not be bothered yet
* commits can probably be merged / rewritten / ..., I split as much as possible to keep the details for now
* the last commit commit message is pretty much MIA, and it probably requires additional tests for edge cases around shebangs / ...

### CODE REVIEW

**Comment by xmo-odoo** [comment]
It does look pretty tho (aside from the precommit step):
```
__ summary __
  py34: commands succeeded
  py35: commands succeeded
  py36: commands succeeded
  py37: commands succeeded
  py38: commands succeeded
  py39: commands succeeded
  py310: commands succeeded
ERROR:   pre-commit: commands failed
tox  76.22s user 8.26s system 116% cpu 1:12.66 total
```

**Comment by peterjc** [comment]
Good work - I've already cherry-picked the test case fixes, a rebase ought to be trivial.

**Comment by peterjc** [comment]
Note the pytest part of this PR would replace #14.

I'm tempted to try to pull out the #38/#49 separately though...

**Comment by peterjc** [comment]
I've split off the offset fix as #52, and suggest merging that - and then rebasing this (#51) which should just be the pytest work.

**Comment by xmo-odoo** [comment]
> Note the pytest part of this PR would replace #14.
> 
> I'm tempted to try to pull out the #38/#49 separately though...

Sure would make sense, as I said this is really here so it can be played with a bit and so I didn't forget things overnight.

The first commit can be left here, the next one (checking the embedded expected outputs) put in a separate PR

> Good work - I've already cherry-picked the test case fixes, a rebase ought to be trivial.

Ok so I can drop the fixes to the test cases as well?

**Comment by xmo-odoo** [comment]
@peterjc saw you enabled the GH actions (because they failed as they could not find the run script anymore).

Updated the yaml to run using tox directly via [tox-gh-actions](https://github.com/ymyzk/tox-gh-actions), which seems to work *really well*, just worked out of the box following the instructions (and also looking at what the pygithub project does as I remembered it also uses tox for testing). Only issue was that I tried enabling 3.4 and apparently it's present in the manifest but not available in the ubuntu-latest image thing, so left it out in the end.

**Comment by peterjc** [comment]
We don't have to worry about Python 3.4 support anymore, even Python 3.6 is on its way out.

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -3,54 +3,20 @@ name: test
...
-        python-version: 3.8
-    - name: Lint code
-      run: |
-        if grep -P '\t' *.py *.rst; then
-            echo 'Tabs are bad, please use four spaces.';
-            false;
-        fi
-        if grep -n -r '[[:blank:]]$' *.py *.rst; then
-            echo 'Please remove trailing whitespace.'; false;
-        fi
```
Comment (peterjc) [inline .github/workflows/test.yml:22]: I should have removed those checks a while ago, and just called pre-commit

**Comment by peterjc** [comment]
I've just merged a tweaked version #53 (leaving out the tox work), and earlier today in 14559893b1f844c177c62d129b7a1809a9b45007 dropped the really elderly Python versions to require Python 3.6 onwards.

I'm going to update this branch / PR now, leaving just the tox work.

**Comment by peterjc** [comment]
OK, I *think* all your valuable contributions have now been applied, except what is left here now - using Tox. However, I doubt I would actually use the Tox setup personally, and we don't need it for GitHub actions.

I have used Tox before, and found it most valuable for testing with multiple versions of Python - which I used to have under $HOME self compiled. However, since then I've adopted conda for dependencies, and that does not seem to make it trivial to have multiple versions of Python installed at the same time (available as binary names ``python3.6`` etc).

Perhaps I am missing something? How are your multiple Python versions setup?

**Comment by xmo-odoo** [comment]
> Perhaps I am missing something? How are your multiple Python versions setup?

I just have a bunch of them installed and activated via `pyenv`.

tox is convenient then, because as long as the relevant pythons are available (and there's a globally installed tox) it handles creating the venvs, setting up the bits and running the tasks, without having to manually create virtualenvs per version of python, or install dependencies globally.

An other bit I find useful is it avoids having to find out how to test the package, if there's a `tox.ini` you know you can just `tox` to run everything, or `tox -l` to see what teh targets are and if the project has an interesting setup (e.g. the `pre-commit` I'd defined here) then maybe run a single environment to get started (e.g. usually to fix one thing you probably want to run just one of the envs).
